File: masterClass/js/challenges/areThereDuplicates.js
/* areThereDuplicates accepts a variable number of arguments, checks whether there are any dupes among the args. Use frequency counter pattern or multiple pointers pattern.

areThereDuplicates(1, 2, 3) -> false
areThereDuplicates(1, 2, 2) -> true
areThereDuplicates('a', 'b', 'c', 'a') -> true

Time: O(n)
Space: O(n)

BONUS:
Time O(log n)
Space O(1)

*/

const areThereDuplicates = function() {
  const memo = {};
  const args = [...arguments];
  for (const argument of args) {
    if (!memo[argument]) memo[argument] = true;
    else return true;
  }
  return false;
};

areThereDuplicates('a', 'b', 'c', 'a');

File: masterClass/py/problem_patterns/challenges/are_there_duplicates.py
def are_there_duplicates(*args):
    memo = {}
    for arg in args:
        if memo.get(arg):
            return True
        memo[arg] = True
    return False

print(are_there_duplicates(1, 2, 3)) # False
print(are_there_duplicates(1, 2, 2)) # True
print(are_there_duplicates("a", "b", "c", "a")) # True